core/src/main/scala/kafka/server/ListOffsetsRequestInstrumentation.scala

```diff
@@ -0,0 +1,133 @@
+  private val earliestTimestampHist: Histogram        = newHistogram(partitionsPerRequest, biased = true,           Map(listBy -> EARLIEST))
+  private val latestTimestampMeter: Meter             = newMeter(partitionRequestRate, eventType, TimeUnit.SECONDS, Map(listBy -> LATEST))
+  private val latestTimestampHist: Histogram          = newHistogram(partitionsPerRequest, biased = true,           Map(listBy -> LATEST))
+  private val liEarliestLocalTimestampMeter: Meter    = newMeter(partitionRequestRate, eventType, TimeUnit.SECONDS, Map(listBy -> LI_EARLIEST_LOCAL))
+  private val liEarliestLocalTimestampHist: Histogram = newHistogram(partitionsPerRequest, biased = true,           Map(listBy -> LI_EARLIEST_LOCAL))
+  private val maxTimestampMeter: Meter                = newMeter(partitionRequestRate, eventType, TimeUnit.SECONDS, Map(listBy -> MAX))
+  private val maxTimestampHist: Histogram             = newHistogram(partitionsPerRequest, biased = true,           Map(listBy -> MAX))
+  private val byTimestampMeter: Meter                 = newMeter(partitionRequestRate, eventType, TimeUnit.SECONDS, Map(listBy -> BY_TIMESTAMP))
+  private val byTimestampHist: Histogram              = newHistogram(partitionsPerRequest, biased = true,           Map(listBy -> BY_TIMESTAMP))
+
+
+  // The object would be periodically dumped to log and cleared on kafka-server wrapper
+  var listOffsetsByTimestampApiClientUsers: mutable.Map[String, concurrent.Map[String, AtomicInteger]] = _
+  snapshotAndResetListOffsetByTimeStampApiUsers()
+
+  /**
+   * A helper method for the external wrapper to obtain the tracked requesters and refresh the tracking map
+   */
+  def snapshotAndResetListOffsetByTimeStampApiUsers(): mutable.Map[String, concurrent.Map[String, AtomicInteger]] = {
+    val old = listOffsetsByTimestampApiClientUsers
+    listOffsetsByTimestampApiClientUsers = mutable.Map()
+    old
+  }
+
+  def logUsage(principal: KafkaPrincipal, topic: ListOffsetsTopic): Unit = {
+    var earliestCnt = 0;
+    var latestCnt = 0;
+    var liEarliestLocalCnt = 0;
+    var maxCnt = 0;
+    var unknownCnt = 0;
+    var byTimestampCnt = 0;
+    topic.partitions().asScala.foreach { partition =>
+      partition.timestamp() match {
+        // special types like EARLIEST are constants < 0
+        case ListOffsetsRequest.EARLIEST_TIMESTAMP => earliestCnt += 1
+        case ListOffsetsRequest.LATEST_TIMESTAMP => latestCnt += 1
+        case ListOffsetsRequest.LI_EARLIEST_LOCAL_TIMESTAMP => liEarliestLocalCnt += 1
+        case ListOffsetsRequest.MAX_TIMESTAMP => maxCnt += 1
+        // Negative, not by actual timestamp, but also not yet defined constant type
+        case t if t < 0 => unknownCnt += 1
+        // When > 0, it's specifying search by an actual timestamp
+        case t if t >= 0 => byTimestampCnt += 1
+      }
+    }
+
+    // Mark partitions requested per second for each type
+    earliestTimestampMeter.mark(earliestCnt)
```

Comment:
why not mark them in the case XXX block?

Reply:
Ah :) 

Actually that was the previous version.  But then the requirement comes out that "we want what's the number of partitions in a request", which means we need gather this aggregate count anyway.

Since we already calculated the count i just mark it here.  But you're right, it's the same as `.mark()` in the case XXX.